core/vm/eof.go

```diff
@@ -0,0 +1,340 @@
+func readEOF1Header(code []byte) EOF1Header {
+	var (
+		header       EOF1Header
+		i            = sectionHeaderStart
+		codeSections = uint16(1)
+	)
+	// Try to read type section.
+	if code[i] == byte(kindType) {
+		size, _ := readSectionSize(code, i+1)
+		header.typeSize = size
+		codeSections = size / 2
+		i += 3
+	}
+	i += 1
+	// Read code sections.
+	for j := 0; j < int(codeSections); j++ {
+		size := binary.BigEndian.Uint16(code[i+3*j : i+3*j+2])
+		header.codeSize = append(header.codeSize, size)
+	}
+	i += int(2 * codeSections)
+	// Try to read data section.
+	if code[i] == byte(kindData) {
+		header.dataSize = binary.BigEndian.Uint16(code[i+1 : i+3])
+	}
+	return header
+}
+
+// validateInstructions checks that there're no undefined instructions and code ends with a terminating instruction
+func validateInstructions(code []byte, header *EOF1Header, jumpTable *JumpTable) error {
+	var (
+		i        = 0
+		analysis = codeBitmap(code)
+		opcode   OpCode
+	)
+	for i < len(code) {
+		switch opcode = OpCode(code[i]); {
+		case jumpTable[opcode].undefined:
+			return fmt.Errorf("%v: %v", ErrEOF1UndefinedInstruction, opcode)
+		case opcode >= PUSH1 && opcode <= PUSH32:
+			i += int(opcode) - int(PUSH1) + 2
+			continue // todo make sure this actually continues
+		case opcode == RJUMP || opcode == RJUMPI:
+			var arg int16
+			// Read immediate argument.
+			if err := binary.Read(bytes.NewReader(code[i+1:]), binary.BigEndian, &arg); err != nil {
+				return ErrEOF1InvalidRelativeOffset
+			}
+			// Calculate relative target.
+			pos := i + 3 + int(arg)
+
+			// Check if offset points to out-of-bounds location.
+			if pos < 0 || pos >= len(code) {
+				return ErrEOF1InvalidRelativeOffset
+			}
+			// Check if offset points to non-code segment.
+			// TODO(matt): include CALLF and RJUMPs in analysis.
+			if !analysis.codeSegment(uint64(pos)) {
+				return ErrEOF1InvalidRelativeOffset
+			}
+			i += 3
+		case opcode == CALLF:
+			var arg int16
+			// Read immediate argument.
+			if err := binary.Read(bytes.NewReader(code[i+1:]), binary.BigEndian, &arg); err != nil {
+				return fmt.Errorf("%v: %v", ErrEOF1InvalidCallfSection, err)
+			}
+			if int(arg) >= len(header.codeSize) {
+				return fmt.Errorf("%v: want section %v, but only have %d sections", ErrEOF1InvalidCallfSection, arg, len(header.codeSize))
+			}
+			i += 3
+		default:
+			i++
+		}
+	}
+	if !opcode.isTerminating() {
+		return fmt.Errorf("%v: %v", ErrEOF1TerminatingInstructionMissing, opcode)
+	}
+	return nil
+}
+
+// readSectionSize returns the size of the section at the offset i.
+func readSectionSize(code []byte, i int) (uint16, error) {
+	if len(code) < i+2 {
+		return 0, fmt.Errorf("section size missing")
+	}
+	return binary.BigEndian.Uint16(code[i : i+2]), nil
+}
+
+type Annotation struct {
+	input  uint8
+	output uint8
+}
+
+type EOF1Container struct {
+	header EOF1Header
+	types  []Annotation // type defs
+	code   []uint64     // code start offsets
```

Comment:
I find this very confusing. If this is indeed _code_, then let it be named `code`. If this is "code statr offsets", then name it `codeOffsets` or something